### ISSUE TITLE
Open-mSupply Global Prefs

### ISSUE THREAD

**Issue by jmbrunskill** [msupply-foundation/open-msupply#5644]
## Is your feature request related to a problem? Please describe 👀

Global prefs are a useful feature to allow system admins to control how the system operates nationally (store prefs do a similar thing at a store level)

>Side Note: Before adding a global prefes, we should consider what makes a good global vs store pref, and/or if a plugin might be more appropriate for each use cases.

We need the ability to easily create and manage store prefs for Open mSupply Specific stores.
We also need to consider how, or if it interacts with mSupply Global Prefs (do they both existing in open-mSupply?)

## Describe the solution you'd like 🎁

From @andreievg ...

I played around a little bit with the translation structure, considering that global preference is stored in a table with key = preference type and value = json. In the use case of the requisition line column visibility some columns might be added via plugin, so they won't be strictly typed, but other should be I think, structure below should allow for it.

```rust

#[derive(Serialize, Deserialize, Debug)]
enum TypedColumnName {
    PreviousStockOnHand,
    PositiveAdjustment,
}
#[derive(Serialize, Deserialize, Debug)]
struct TypedColumn {
    column_name: TypedColumnName,
    is_visible: bool,
}
#[derive(Serialize, Deserialize, Debug)]
struct CustomColumn {
    column_name: String,
    is_visible: bool,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(untagged)]
enum RequisitionColumnVisibility {
    TypedColumn(TypedColumn),
    CustomColumn(CustomColumn),
}

fn main() {
    let value = r#"[{
        "column_name": "PreviousStockOnHand", "is_visible": true },
         {"column_name": "PositiveAdjustment", "is_visible": false },
         {"column_name": "SomeCustomColumn", "is_visible": true }
    ]"#;

    println!(
        "{:#?}",
        serde_json::from_str::<Vec<RequisitionColumnVisibility>>(value)
    );

    println!(
        "{}",
        serde_json::to_string_pretty(&vec![RequisitionColumnVisibility::TypedColumn(
            TypedColumn {
                column_name: TypedColumnName::PositiveAdjustment,
                is_visible: true
            }
        )])
        .unwrap()
    );
}
```

Above can go all the way through to graphql and typescript layer i think.

I would also imagine that we access each preference type separately, and have a seperate method for each serialisation and mapping, although I played around with some generic preference types, not sure how they would work across all of the layers:

```rust

table! {
    glob_pref (id) {
        id -> Text,
        value -> Text,
    }
}

trait GlobalPref: Default + DeserializeOwned {
    fn key() -> &'static str;

    fn load(connection: &StorageConnection) -> Result<Self, RepositoryError> {
        use glob_pref::dsl;

        let text_pref: Option<String> = dsl::glob_pref
            .filter(dsl::id.eq(Self::key()))
            .select(dsl::value)
            .first(connection.lock().connection())
            .optional()?;

        let Some(text_pref) = text_pref else {
            return Ok(Default::default());
        };

        let parsed = serde_json::from_str(&text_pref).unwrap();

        Ok(parsed)
    }
}

#[derive(Default, Deserialize)]
struct CustomTranslation(serde_json::Value);

impl GlobalPref for CustomTranslation {
    fn key() -> &'static str {
        "custom_translations"
    }
}

#[derive(Default, Deserialize)]
struct CustomTranslation(serde_json::Value);
```

### Moneyworks Jobcode 🧰

OMS:DDSP

**Comment by jmbrunskill** [comment]
Needed for https://github.com/msupply-foundation/open-msupply/issues/5138#issuecomment-2461246641
and https://github.com/msupply-foundation/open-msupply/issues/5061

**Comment by jmbrunskill** [comment]
@Chris-Petty I think this is going to need a bit of UI for editng prefs too.

Can you add another issue for the UI side of this?

**Comment by Chris-Petty** [comment]
Sure! Just confirming things with Andrei too

**Comment by Chris-Petty** [comment]
TL;DR: We really need to lock down the use cases of each pref we introduce and if they're store or global prefs. The prefs I've had highlighted to me so far aren't either, but non-sync datafile prefs. If they did sync, both should be store prefs IMO. If they truly are global prefs, we should implement config UI in OMS and have some duplication of setup in the seemingly rare case OG actually treats them as global synced prefs. This also moves us further down the path to eliminating the OG central server.

Going to flag here some background that can be summarised as OG prefs are a hot mess of different things over different decades 😉. We can't assume anything in the "preferences" window is global pref. Some are global and synced, some are for just for the open datafile, shared across the stores of the datafile, and not synced at all. They can be in the `pref` table or the `pref_blob` table. The former has _some_ rules for data structure, while the latter is pretty loose. For instance

```4d
	If (cb_placeholder_lines=1)
		<>ipref_inv_entry:=(<>ipref_inv_entry ?+ 26)
	End if 
	
	If (cb_margin_pref=1)
		<>ipref_inv_entry:=(<>ipref_inv_entry ?+ 27)
	End if 
	
	If (cb_direction_entry=1)
		<>ipref_inv_entry:=(<>ipref_inv_entry ?+ 28)
	End if 
	
	If (cb_mono_spaced_font=1)
		<>ipref_inv_entry:=(<>ipref_inv_entry ?+ 29)
	End if 
```

Whether or not to display directions isn't currently a Global Preference, it is a datafile preference so does not sync. It's stored in a `pref` record (I was surprised, figured it'd be old and in `pref_blob`!). While the pref is technically in an object key:value structure, this pref record is essentially `pref.data.data: i32`, where the integer assigned is acting as a bit array. "Item directions shown in dispensing" is bit 28. Also stored in the bit array is prefs for whether or not to use placeholders, backorders, some kind of margins, mono spaced fonts (for somewhere, IDK). It's a very early structure probably predating sync itself controlling a handful of `transact` related preferences, so fairly broad!

This makes me more confident we should just implement global prefs in OMS and only implement prefs that we're absolutely sure we want, least we're changing behaviour in OG too. And again as mentioned elsewhere by Drei, be vary conscious of what should be a store or global pref. Generally I'd say it's safer to assume store level for finer grain control, with being able to set it for all stores being essentially global if that's necessary. 

As an example, the directions entry pref defaults to on in new datafiles. You may want to disable it if a facility (i.e. where the datafile/site is physically located and operates) has no label printers and no plan to get them. As a Global pref you loose that. With store pref you'd probably have enough specific control to allow that to be disabled.

n.b. I don't think we should have datafile specific prefs. If anything, it should be a site specific pref so it can be backed up via sync. Store prefs are finer grained, already exist and sync, so may be somewhat redundant having site prefs.

--- 

But wait there's more! In the acceptance criteria we're looking for `"allowWardSelectionOnPrescriptions"` as the example of a global pref to start with, however:

1. It's not currently a global pref. It doesn't sync when updated. That is to say, it's a datafile pref.
2. You'd want store level control for this I'd have thought. A hospital may have multiple stores, some for inpatients and some for outpatients. An inpatient pharmacy store may want to specify the ward of the patient, while an outpatient pharmacy store would not want that functionality.

---

Final n.b. (hopefully): Initialisation has a bit of a everything and the kitchen sync approach to prefs. It just syncs basically all of them. That includes most of the central server's datafile prefs. This can be confusing, case while they're in the sync buffer of remote sites, they do not get any updates if that pref is changed on the central server.

**Comment by andreievg** [comment]
Comment from refinement: keep existing store prefs for now and have two end points and two services. And create an issue with 'pathway' for the future and communicate this change and variation of patterns

**Comment by Chris-Petty** [comment]
Closing as part of https://github.com/msupply-foundation/open-msupply/issues/6836
